policy-controller/k8s/index/src/pod.rs

```diff
@@ -0,0 +1,367 @@
+        )
+    )]
+    pub(crate) fn apply_pod(&mut self, pod: k8s::Pod) -> Result<()> {
+        let Namespace {
+            default_allow,
+            ref mut pods,
+            ref mut servers,
+            ..
+        } = self
+            .namespaces
+            .get_or_default(pod.namespace().expect("namespace must be set"));
+
+        let default_allow = *default_allow;
+        let allows = self.default_allows.clone();
+        let mk_default_allow =
+            move |da: Option<DefaultAllow>| allows.get(da.unwrap_or(default_allow));
+
+        pods.apply(
+            pod,
+            &mut self.nodes,
+            servers,
+            &mut self.lookups,
+            mk_default_allow,
+        )
+    }
+
+    #[instrument(
+        skip(self, pod),
+        fields(
+            ns = ?pod.metadata.namespace,
+            name = ?pod.metadata.name,
+        )
+    )]
+    pub(crate) fn delete_pod(&mut self, pod: k8s::Pod) -> Result<()> {
+        let ns_name = pod.namespace().expect("namespace must be set");
+        let pod_name = pod.name();
+        self.rm_pod(ns_name.as_str(), pod_name.as_str())
+    }
+
+    fn rm_pod(&mut self, ns: &str, pod: &str) -> Result<()> {
+        if self.nodes.clear_pending_pod(ns, pod) {
+            // If the pod was pending that it can't be in the main index.
+            debug!("Cleared pending pod");
+            return Ok(());
+        }
+
+        self.namespaces
+            .index
+            .get_mut(ns)
+            .ok_or_else(|| anyhow!("namespace {} doesn't exist", ns))?
+            .pods
+            .index
+            .remove(pod)
+            .ok_or_else(|| anyhow!("pod {} doesn't exist", pod))?;
+
+        self.lookups.unset(&ns, &pod)?;
+
+        debug!("Removed pod");
+
+        Ok(())
+    }
+
+    #[instrument(skip(self, pods))]
+    pub(crate) fn reset_pods(&mut self, pods: Vec<k8s::Pod>) -> Result<()> {
+        self.nodes.clear_pending_pods();
+
+        let mut prior_pods = self
+            .namespaces
+            .iter()
+            .map(|(name, ns)| {
+                let pods = ns.pods.index.keys().cloned().collect::<HashSet<_>>();
+                (name.clone(), pods)
+            })
+            .collect::<HashMap<_, _>>();
+
+        let mut result = Ok(());
+        for pod in pods.into_iter() {
+            let ns_name = pod.namespace().unwrap();
+            if let Some(ns) = prior_pods.get_mut(ns_name.as_str()) {
+                ns.remove(pod.name().as_str());
+            }
+
+            if let Err(error) = self.apply_pod(pod) {
+                result = Err(error);
+            }
+        }
+
+        for (ns, pods) in prior_pods.into_iter() {
+            for pod in pods.into_iter() {
+                if let Err(error) = self.rm_pod(ns.as_str(), pod.as_str()) {
+                    result = Err(error);
+                }
+            }
+        }
+
+        result
+    }
+}
+
+// === impl PodIndex ===
+
+impl PodIndex {
+    fn apply(
```

Comment:
nit: the name `apply` is generic enough that it might be worth having a comment explaining wat, exactly, this method actually does...

Reply:
it's not that ambiguous in the context of kubernetes -- it means create or update (as in `kubectl apply`)